pkg/codegen/java/gen.go

```diff
@@ -599,6 +608,15 @@ func (pt *plainType) genInputType(ctx *classFileContext) error {
 	for propIndex, prop := range pt.properties {
 		propType := propTypes[propIndex]
 		fieldName := names.Ident(pt.mod.propertyName(prop)).AsProperty().Field()
+
+		if prop.IsRequired() && prop.DefaultValue == nil && prop.ConstValue == nil {
+			fprintf(w, "            if ($.%s == null) {\n", fieldName)
+			fprintf(w, "                throw new %s(\"%s\", \"%s\");\n",
+				ctx.ref(names.PulumiMissingRequiredPropertyException), pt.name, fieldName)
```

Comment:
Are we going to need to first ship the exception in the core Java SDK, and then somehow have the codegen indicate that version of the core Java SDK is now required (since it depends on this new exception being there)? I'm not familiar enough with Java on how to indicate that requirement.

Reply:
Yes we do.

Reply:
I thought it should be fine because if it is merged and tag a release, a new sdk is published that includes the new exception type, sometime later providers team will pick up the newer sdk-gen from pulumi-java and start using that new type. This is assuming that sdk-gen references latest pulumi-java sdk

Reply:
sdk-gen dictates the version of the core library to use. So we need to release the type first, then make sure sdk-gen is setting that version as the required version when emiting the provider sdks.

Reply:
i.e. "This is assuming that sdk-gen references latest pulumi-java sdk" is the assumption that doesn't hold. e.g. [this line](https://github.com/pulumi/pulumi-aws/blob/master/sdk/java/build.gradle#L46) which will be set from in https://github.com/pulumi/pulumi-java/blob/main/pkg/codegen/java/templates_gradle.go#L87 somehow.

Reply:
> i.e. "This is assuming that sdk-gen references latest pulumi-java sdk" is the assumption that doesn't hold.

The `pulumi-java-gen` CLI used to generate java SDKs takes the version of the pulumi-java main sdk as input but it defaults to the version of the `pulumi-java-gen` itself, see [here](https://github.com/pulumi/pulumi-java/blob/main/pkg/cmd/pulumi-java-gen/command.go#L91-L92):
```go
cmd.Flags().StringVar(&javaSdkVersionArg, "sdk", version.Version,
	"com.pulumi:pulumi version to depend on in the generated code")
```
Doesn't that mean if we tag 0.9.x which publishes both SDK and generator, that the generated SDKs will be using 0.9.x as well? 🤔

Reply:
What about from the Pulumi cli where we call the gen methods directly?

Reply:
Tested what Pulumi CLI does when you run:
```
pulumi package gen-sdk schema.json --language java --out sdk
```
and it generates just java source files without a `gradle` project file 😱 so basically gen-sdk isn't usable with java projects

Reply:
🤦‍♂️ alright, well we'll need to look into that at some point but means nothing to worry about for this PR